crates/pico_macros/src/memo.rs

```diff
@@ -0,0 +1,136 @@
+use std::hash::DefaultHasher;
+use std::hash::Hash;
+use std::hash::Hasher;
+
+use darling::{ast::NestedMeta, FromMeta};
+use proc_macro::TokenStream;
+use quote::quote;
+use quote::ToTokens;
+use syn::Signature;
+use syn::{
+    parse_macro_input, parse_quote, Error, FnArg, GenericParam, ItemFn, PatType, ReturnType,
+};
+
+#[derive(Debug, FromMeta)]
+struct MemoArgs {
+    #[darling(default)]
+    reference: bool,
+}
+
+pub(crate) fn memo(args: TokenStream, item: TokenStream) -> TokenStream {
+    let ItemFn {
+        sig,
+        vis,
+        block,
+        attrs,
+    } = parse_macro_input!(item as ItemFn);
+
+    let fn_hash = hash(&sig);
```

Comment:
Can we detect collisions? (I don't care right now, to be fair)

e.g. if we have a `lazy_static!( I_HAVE_BEEN_SET: false )` which we set to true when the memoized function is called initially. And the we should only store an `inner_fn` under that `fn_hash` when `I_HAVE_BEEN_SET` is false.

Probably worth not doing in v1, I don't think we'll personally ever encounter collisions

Reply:
I believe we can find something smarter than just a fn body hash, kind of immutable map that we build at compile time. Just haven't figured out it yet.